### DISCUSSION TITLE
Show _some_ data automatically in 3D panel

### DISCUSSION THREAD

**Discussion by jtbandes** [foxglove/community#100] [Q&A]
When using a new, blank 3D panel and connecting to any data source, the panel doesn't display anything until the user configures it in settings. This causes user confusion (e.g. https://github.com/foxglove/ws-protocol/issues/278#issuecomment-1324544395). We should at least display _some_ topics by default.

Originally, the default behavior of the new 3d panel was to subscribe to **all** available topics, but that was deemed too chaotic and data-intensive. Hiding only camera images was proposed, but we settled on hiding everything by default:
- https://github.com/foxglove/studio/issues/3643
- https://github.com/foxglove/studio/issues/3692

We've discussed displaying a **grid** by default (and some technical limitations that make this harder than it sounds):
- https://github.com/foxglove/studio/issues/4312
- https://github.com/foxglove/studio/issues/3795

Separately from the grid, though, it would still be better to show some actual user data without requiring the user to discover and understand panel settings.

Of course, if we choose to show some but not all user data by default, it might cause confusion when users don't understand some data is automatically hidden, and just assume that some of their data doesn't work. So we may need to improve settings discoverability as well. (cc @2metres @ocin-rye @esthersweon)

Copying my ideas from https://github.com/foxglove/studio/issues/3643#issuecomment-1164941057:

> Here are some ideas for heuristics if we decide that showing all topics is too much.
> - Prioritize topics by data size, if we know the size, otherwise sort by schema in order to estimate how expensive it is to show topics.
>   - e.g. if we have to choose between a Pose or a Marker and an Image message, we should drop the Image first since it's likely to be much more expensive to visualize.
>   - Then show at most 20 topics by default in this priority order.
> - If we didn't turn on all topics by default, show the user a message/popup saying something to the effect of "some topics are hidden, open settings sidebar to customize".
> - Keep in mind that the same layout may be used with different data sources over time, or a single live data source that gains more topics. So we have to respond dynamically to new topics coming online. This might be tricky since we don't want to store the list of "seen topics" in the layout. We might need to keep careful track of which topics the user has manually toggled on/off, and only dynamically decide what to do for topics that have never been manually toggled. [related: foxglove/studio#4809]
> - Or just don't do any of this, because it sounds complicated and fragile

And some more ideas:
- Show all topics of certain types by default (markers, poses, grids, etc.) but not cameras, **unless** there are _only_ cameras available

**Comment by esthersweon** [comment]
I like the idea of showing all topics by default – it makes it more obvious that all the user's data is there, and avoids the need for a panel-specific info popup (which I think would make the UI busier / more confusing). 

I think the cog icon in the top right of each panel is a relatively obvious way to configure the panel settings. Maybe we could highlight it on first load with a pulsing animation of some sort?

As for switching data sources, I wonder if we can have this default-all behavior only for the first time a given 3D panel is loaded, to simplify things. (Do we currently prune topics saved in the layout that don't exist in the data source? If not, I think it's fine to keep that behavior - esp. if we have a way to direct users to the panel settings more clearly.)

**Reply by jtbandes** [reply]
We do highlight the button like this the first time the panel loads (it's permanently dismissed when you click the X or the ⚙️ itself):
<img width="264" alt="image" src="https://user-images.githubusercontent.com/14237/208744954-ace30121-2af8-4bd5-9ea8-649a2f553ea8.png">

**Comment by jtbandes** [comment]
Another example of user confusion: https://github.com/foxglove/studio/issues/4732#issuecomment-1371991820

**Reply by defunctzombie** [reply]
I will continue to say that a simple message in the panel indicating the user has no topics enabled would be easy to implement and require no heuristics or guessing.

**Reply by jtbandes** [reply]
Internal tracking ticket: https://linear.app/foxglove/issue/FG-1098

**Comment by egradman** [comment]
As a new user who just spent an entire day painstakingly crafting and re-crafting `SceneUpdate` messages to get my meshes to appear—only to discover I merely had to click the deeply nested "show stuff" button—I would like to enthusiastically +1 this issue.  My thinking was "my `FrameTransforms` are appearing automatically, so why wouldn't my `SceneEntity`s appear automatically too?"

**Comment by defunctzombie** [comment]
I'm gonna close this. We aren't gonna show any topics by default because its a guess at what topics and loading topics can have a cost and its not clear why some topics would be shown and not others. I still think a grid should be shown by default but as for other topics no plans to make that happen.
